packages/react-router/modules/matchPath.js

```diff
@@ -25,7 +79,7 @@ function compilePath(path, options) {
 /**
  * Public API for matching a URL pathname to a path.
  */
-function matchPath(pathname, options = {}) {
+function matchPath(pathname, options = {}, parentMatch = null) {
```

Comment:
I think that using the full parent match is overkill. using just the pathname would be enough.

The pathname allows us to extract the necessary params just as well.

Right now, you do:

`match = regexp.exec("/user/1234/" + "profile")` + mixing in the parent params into the `match` object.

You could instead do just one step:

`match = regexp.exec("/user/:id/" + "profile")` and no manual mixing in of the params necessary.

Reply:
Not sure about this point. I will think about this in the next minutes.

Reply:
okay, did it!

Reply:
The point is the use of `pathname` over `url`. 

In my scenario, you have:

relative path: `./profile`
parentMatch.url: `/user/1234` 
parentMatch.pathname: `/user/:id` 
parentMatch.params: `{ id: "1234 }` 

And instead of using the pathname + relative part to put through `path-to-regex`, you concat  url + relative path.

That means you also have to combine parentMatch.params with the `params` object you assembled from `path-to-regex`.

Or alternatively:

```js
const path = resolvePath(path, parentMatch && parentMatch.url);
const match = regexp.exec(path);
// skipping some details here.
const matchParams = keys.reduce((params, key, index) => {
  params[key.name] = values[index];
  return params;
}, {});
const parentParams = (parentMatch && parentMatch.params) || {};
const params = absolute ? matchParams : { ...parentParams, ...matchParams };
```

vs.

```js
const path = resolvePath(path, parentMatch && parentMatch.pathname);
const match = regexp.exec(path);
// skipping some details here.
const params = keys.reduce((params, key, index) => {
  params[key.name] = values[index];
  return params;
}, {});
```

Reply:
Okay, I changed it to your enhancement. With one little difference: it is `parentMatch.path` and not `parentMatch.pathname` :wink:

Reply:
I always swap location.pathname with match.path.

You still have to remove the now redundant code btw.

Reply:
I am not sure about your last sentence: what redundant code do you mean?

Reply:
```js
const parentParams = (parentMatch && parentMatch.params) || {};
const params = absolute ? matchParams : { ...parentParams, ...matchParams };
```

Reply:
I am not sure from where you get this, but this is not part of the latest commits.

Reply:
Huh. Sorry.

I was tired and probably checked the wrong commit by accident.

Reply:
No problem, this happens sometimes. I was just confused about if I have missed something. Probably I was tired too.